### ISSUE TITLE
Writing string vector with mixed encodings causes corrupted table contents

### ISSUE THREAD

**Issue by aryoda** [r-dbi/odbc#172]
I am investigating encoding problems with `odbc` and MS SQL Server (see #56) and can reproduce strange behaviour on **Ubuntu using MySQL (5.7.21 with innodb)** so it is easier to debug without Windows and MS SQL Server.

# Observations

If you write an R character vector with mixed encodings (UTF-8, latin1, ...) into a DB table that was created manually using the correct encoding (UTF8mb4 but also tested with UTF8)...

1. the content of the table is displayed incorrectly (e. g. using a DB frontend like MySQL Workbench 6.5.2.0)
2. the content is read back incorrectly
3. while entering a UTF-8 text directly into the table does display the correct string in the MySQL Workbench

**Side note**: If `dbWriteTable` creates the DB table the data can be read back to R correctly (= same bytes) but is still displayed incorrectly in the DB table!

A related issue may be: #143

# Steps to reproduce the problem

Install MySQL using docker (if not already present): `docker pull mysql`

Start MySQL with:

```
docker run --name DBImysql -e MYSQL_ROOT_PASSWORD=my_secret -e MYSQL_DATABASE=testdb -p 30001:3306 -d mysql
```

Then use this R script:

```
library(DBI)
# library(odbc)   # devtools::install_github("r-dbi/odbc")

# test data taken from package DBItest:
# https://github.com/r-dbi/DBItest/blob/master/R/utf8.R
text_cyrillic <- "\u041a\u0438\u0440\u0438\u043b\u043b"
text_latin <- "M\u00fcller"
text_latin_encoded <- iconv(text_latin, from = "UTF-8", to = "latin1")
text_chinese <- "\u6211\u662f\u8c01"
text_ascii <- iconv("ASCII", to = "ASCII")

mixed_enc <- c(text_cyrillic, text_latin, text_latin_encoded, text_chinese, text_ascii)

data <- data.frame(ID = 1:length(mixed_enc),
                   mixed_enc = mixed_enc,
                   enc = Encoding(mixed_enc),
                   stringsAsFactors = FALSE)

#   ID mixed_enc     enc
# 1  1    Кирилл   UTF-8
# 2  2    Müller   UTF-8
# 3  3    Müller  latin1
# 4  4    我是谁   UTF-8
# 5  5     ASCII unknown

# ensure pre-condition
stopifnot( Encoding(mixed_enc) == c("UTF-8", "UTF-8", "latin1", "UTF-8", "unknown"))

con <- dbConnect(odbc::odbc()
                 , .connection_string = "driver=MySQL;server=127.0.0.1;port=30001;user=root;password=my_secret;database=testdb;option=3;protocol=tcp"
                 , encoding = "UTF-8") # utf8mb4 didn't work neither (if encoding should mean "connection character set")

table_name <- "encoding_test"
if (dbExistsTable(con, table_name)) { dbRemoveTable(con, table_name) }
# trying utf8 instead utf8mb4 does not help and is not recommended, see:
# https://medium.com/@adamhooper/in-mysql-never-use-utf8-use-utf8mb4-11761243e434
sql <- paste0("CREATE TABLE ", table_name, "(ID int, mixed_enc varchar(255) CHARACTER SET utf8mb4, enc varchar(255))")
dbGetQuery(con, sql)

# nanodbc/nanodbc.cpp:1587: 42000: [MySQL][ODBC 5.1 Driver][mysqld-5.7.21]SET NAMES not allowed by driver
# dbGetQuery(con, "SET NAMES utf8;") # communicate with the server using the utf8 character set

dbWriteTable(con, table_name, data, overwrite = FALSE, append = TRUE)

# dbWriteTable(con, table_name, data, overwrite = TRUE)
# Note: Using overwrite = TRUE creates the wrong CHARACTER SET for the mixed_enc column (causing corrupted characters)
#       but reading back the rows re-creates the same strings again in R (this is still not correct since
#       the database content is corrupted). dbWriteTable creates this table:
# CREATE TABLE `encoding_test` (
#   `ID` int(11) DEFAULT NULL,
#   `mixed_enc` text,
#   `enc` text
# ) ENGINE=InnoDB DEFAULT CHARSET=latin1;

res <- dbReadTable(con, table_name)
dbDisconnect(con)

res$res_enc <- Encoding(res$mixed_enc)

print(res)     # for debugging
#   ID      mixed_enc     enc res_enc
# 1  1   ÐšÐ¸Ñ€Ð¸Ð»Ð»   UTF-8   UTF-8
# 2  2        MÃ¼ller   UTF-8   UTF-8
# 3  3        MÃ¼ller  latin1   UTF-8
# 4  4 æˆ‘æ˜¯è°\u0081   UTF-8   UTF-8
# 5  5          ASCII unknown unknown

# SELECT * FROM testdb.encoding_test;
#
#   ID      mixed_enc     enc
#    1   ÐšÐ¸Ñ€Ð¸Ð»Ð»   UTF-8
#    2        MÃ¼ller   UTF-8
#    3        MÃ¼ller  latin1
#    4 æˆ‘æ˜¯è°\u0081   UTF-8
#    5          ASCII unknown
```

# sessionInfo()

```
R version 3.4.4 (2018-03-15)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 14.04.5 LTS

Matrix products: default
BLAS: /usr/lib/atlas-base/atlas/libblas.so.3.0
LAPACK: /usr/lib/lapack/liblapack.so.3.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] DBI_0.8

loaded via a namespace (and not attached):
 [1] bit_1.1-12      odbc_1.1.5      compiler_3.4.4  hms_0.4.2       tools_3.4.4     yaml_2.1.14     Rcpp_0.12.16   
 [8] bit64_0.9-7     blob_1.1.1      pkgconfig_2.0.1 rlang_0.2.0   

```

# DB config

```
SHOW VARIABLES WHERE Variable_name LIKE 'character\_set\_%' OR Variable_name LIKE 'collation%';

'character_set_client', 'utf8'
'character_set_connection', 'utf8'
'character_set_database', 'latin1'
'character_set_filesystem', 'binary'
'character_set_results', 'utf8'
'character_set_server', 'latin1'
'character_set_system', 'utf8'
'collation_connection', 'utf8_general_ci'
'collation_database', 'latin1_swedish_ci'
'collation_server', 'latin1_swedish_ci'
```

# Links

I could try to tweak many settings:
https://mathiasbynens.be/notes/mysql-utf8mb4

https://dev.mysql.com/doc/refman/5.7/en/charset-mysql.html

**Comment by jimhester** [comment]
I cannot reproduce this, perhaps the issue is your database is in latin1, mine is in utf-8 / utf8mb4

```r
dbGetQuery(con, "SHOW VARIABLES WHERE Variable_name LIKE 'character\\_set\\_%' OR Variable_name LIKE 'collation%'")
#>               Variable_name              Value
#> 1      character_set_client               utf8
#> 2  character_set_connection               utf8
#> 3    character_set_database               utf8
#> 4  character_set_filesystem             binary
#> 5     character_set_results
#> 6      character_set_server            utf8mb4
#> 7      character_set_system               utf8
#> 8      collation_connection    utf8_general_ci
#> 9        collation_database    utf8_general_ci
#> 10         collation_server utf8mb4_unicode_ci
```

**Comment by jimhester** [comment]
Did you try setting an `encoding = "latin1"` in the call to `dbConnect()`?

**Comment by aryoda** [comment]
Works, thanks for looking into it and sorry for wasting your rare time due to my blindness.

It works now and my setup above is not helpful to debug the Microsoft SQL Server problem with encodings (as it was my intention).

Good point at least: It proves that `DBI` + `ODBC` can handle mixed encodings so it could be just a similar config problem with the MS SQL Server like it was the case here.

The solution (as you anticipated) was to set the correct MySQL variable values by adding this code after opening the connection:

```
conf1 <- dbGetQuery(con, "ALTER DATABASE testdb CHARACTER SET = utf8 COLLATE = utf8_unicode_ci;")
conf2 <- dbGetQuery(con, "SET character_set_server = utf8;")
conf3 <- dbGetQuery(con, "SET CHARACTER_SET_CONNECTION = 'utf8';")
conf4 <- dbGetQuery(con, "SET CHARACTER_SET_CLIENT = 'utf8';")
```
Output in R (as expected) and also the direct SQL query in MySQL Workbench shows the same result:

```
#   ID mixed_enc     enc res_enc
# 1  1    Кирилл   UTF-8   UTF-8
# 2  2    Müller   UTF-8   UTF-8
# 3  3    Müller  latin1   UTF-8
# 4  4    我是谁    UTF-8   UTF-8
# 5  5     ASCII unknown unknown
```

Variable values of the working version are (captured from within the connection in R):

```
#               Variable_name           Value
# 1      character_set_client            utf8
# 2  character_set_connection            utf8
# 3    character_set_database            utf8
# 4  character_set_filesystem          binary
# 5     character_set_results                
# 6      character_set_server            utf8
# 7      character_set_system            utf8
# 8      collation_connection utf8_general_ci
# 9        collation_database utf8_unicode_ci
# 10         collation_server utf8_general_ci

```

BTW: When I set the `encoding = "latin1"`argument in `dbConnect()` only the latin1-encoded strings are correct (as it shall be), but utf8 strings are damaged:

```
# > print(res)     # for debugging
# ID mixed_enc     enc res_enc
# 1  1      <NA>   UTF-8 unknown
# 2  2    Müller   UTF-8   UTF-8
# 3  3    Müller  latin1   UTF-8
# 4  4      <NA>   UTF-8 unknown
# 5  5     ASCII unknown unknown
```

**Comment by aryoda** [comment]
I will close this issue and open a new one for MS SQL Server if cannot find a solution and can set-up a reproducible example.

Do you have a Windows client and MS SQL Server as infrastructure to examine this?

**Comment by jimhester** [comment]
Yes, I can test environment against SQLServer on Windows

**Comment by aryoda** [comment]
Perfekt. Perhaps the problem is already solved via #179 :-)

I will verify this in the next days...
